File: elixir/protein-translation/protein_translation.exs
defmodule ProteinTranslation do
  @doc """
  Given an RNA string, return a list of proteins specified by codons, in order.
  """
  @spec of_rna(String.t()) :: {atom, list(String.t())}
  def of_rna(rna) do
    rna
    |> String.to_charlist
    |> Enum.chunk_every(3)
    |> Enum.map(&List.to_string/1)
    |> Enum.map(&of_codon/1)
    |> Enum.reduce_while([], &build_rna/2)
  end

  def build_rna
  def build_rna({:ok, codon}, acc) when codon == "STOP" do {:halt, {:ok, acc}} end
  def build_rna({:ok, codon}, acc) do {:cont, acc ++ [codon]} end
  def build_rna({:error, _}, _) do {:halt, {:error, "invalid RNA"}} end

  @doc """
  Given a codon, return the corresponding protein

  UGU -> Cysteine
  UGC -> Cysteine
  UUA -> Leucine
  UUG -> Leucine
  AUG -> Methionine
  UUU -> Phenylalanine
  UUC -> Phenylalanine
  UCU -> Serine
  UCC -> Serine
  UCA -> Serine
  UCG -> Serine
  UGG -> Tryptophan
  UAU -> Tyrosine
  UAC -> Tyrosine
  UAA -> STOP
  UAG -> STOP
  UGA -> STOP
  """
  @spec of_codon(String.t()) :: {atom, String.t()}
  def of_codon(codon) do
    codons = %{
      "UGU" => "Cysteine",
      "UGC" => "Cysteine",
      "UUA" => "Leucine",
      "UUG" => "Leucine",
      "AUG" => "Methionine",
      "UUU" => "Phenylalanine",
      "UUC" => "Phenylalanine",
      "UCU" => "Serine",
      "UCC" => "Serine",
      "UCA" => "Serine",
      "UCG" => "Serine",
      "UGG" => "Tryptophan",
      "UAU" => "Tyrosine",
      "UAC" => "Tyrosine",
      "UAA" => "STOP",
      "UAG" => "STOP",
      "UGA" => "STOP",
    }
    case Map.get(codons, codon) do
      nil -> {:error, "invalid codon"}
      protein -> {:ok, protein}
    end
  end
end

File: python/protein-translation/protein_translation.py
CODON_TO_AMINO_ACID = {
    "AUG": "Methionine",
    "UUU": "Phenylalanine",
    "UUC": "Phenylalanine",
    "UUA": "Leucine",
    "UUG": "Leucine",
    "UCU": "Serine",
    "UCC": "Serine",
    "UCA": "Serine",
    "UCG": "Serine",
    "UAU": "Tyrosine",
    "UAC": "Tyrosine",
    "UGU": "Cysteine",
    "UGC": "Cysteine",
    "UGG": "Tryptophan",
    "UAA": "STOP",
    "UAG": "STOP",
    "UGA": "STOP",
}


def proteins(strand):
    chain = []

    for i in range(len(strand) // 3):
        codon = strand[i * 3 : (i * 3) + 3]
        amino_acid = CODON_TO_AMINO_ACID.get(codon, None)
        if amino_acid == "STOP":
            break
        chain.append(amino_acid)

    return list(filter(None, chain))
